climada/util/coordinates.py

```diff
@@ -1057,6 +1057,69 @@ def assign_coordinates(coords, coords_to_assign, distance="euclidean",
                 coords_to_assign, coords[not_assigned_idx_mask], threshold, **kwargs)
     return assigned_idx
 
+def assign_centroids_to_gdf(coord_gdf, centroids, distance='euclidean',
+                        threshold=NEAREST_NEIGHBOR_THRESHOLD,test_crs=True):
+    """Assign to each gdf coordinate point its closest centroids's coordinate.
+    If disatances > threshold in points' distances, -1 is returned.
+    If centroids are in a raster and coordinate point is outside of it ``-1`` is assigned
+
+    Parameters
+    ----------
+    coord_gdf : gpd.GeoDataFrame
+        GeoDataframe with defined latitude/longitude column and crs
+    centroids : Centroids
+        (Hazard) centroids to match (as raster or vector centroids).
+    distance : str, optional
+        Distance to use in case of vector centroids.
+        Possible values are "euclidean", "haversine" and "approx".
+        Default: "euclidean"
+    threshold : float, optional
+        If the distance (in km) to the nearest neighbor exceeds `threshold`,
+        the index `-1` is assigned.
+        Set `threshold` to 0, to disable nearest neighbor matching.
+        Default: ``NEAREST_NEIGHBOR_THRESHOLD`` (100km)
+    test_crs : bool, optional
+        Whether or not to test equal crs from centroids and gdf
+        In case of exp.assign_centroids() the crs is not defined within
+        the GeoDataFrame and the crs match is tested there. Default: True.
+
+    See Also
+    --------
+    climada.util.coordinates.assign_grid_points: method to associate centroids to
+        coordinate points when centroids is a raster
+    climada.util.coordinates.assign_coordinates: method to associate centroids to
+        coordinate points
+    Notes
+    -----
+    The default order of use is:
+
+        1. if centroid raster is defined, assign the closest raster point
+        to each gdf coordinate point.
+        2. if no raster, assign centroids to the nearest neighbor using
+        euclidian metric
+
+    Both cases can introduce innacuracies for coordinates in lat/lon
+    coordinates as distances in degrees differ from distances in meters
+    on the Earth surface, in particular for higher latitude and distances
+    larger than 100km. If more accuracy is needed, please use 'haversine'
+    distance metric. This however is slower for (quasi-)gridded data,
+    and works only for non-gridded data.
+    """
+
+    if test_crs and (not equal_crs(coord_gdf.crs, centroids.crs)):
+        raise ValueError('Set hazard and GeoDataFrame to same CRS first!')
```

Comment:
```suggestion
    try:
        if not equal_crs(coord_gdf.crs, centroids.crs):
            raise ValueError('Set hazard and GeoDataFrame to same CRS first!')
    except AttributeError:
        pass
```

Reply:
I assume this is a suggestion to avoid the test_crs=True flag? 
Why is this better than the test-crs flag? I thought such pass statements should be avoided according to the [climada guide](https://github.com/CLIMADA-project/climada_python/blob/feature/assign_centroids_impact/doc/guide/Guide_PythonDos-n-Donts.ipynb), but I'm happy to learn :)

Reply:
> ... such pass statements should be avoided ...

Like in " 2. Do not catch exception if you do not handle them."?
 But you totally handle it: if there is no crs in coord_gdf there is nothing to compare and no reason to fail.

Why I don't like the `test_crs` flag?
 Because one is not free to give it a value. There doesn't seem to be a situation where you could set it to `True` _or_ `False` deliberately. The argument is strongly coupled to the `coord_gdf` argument. It increases complexity unnecessarily. It's worse than `pass`ing.

Reply:
_However_ - The `try`/`pass` approach is due to the `crs` being a property (I think) and the canonical way to check for it being there. If there was another way like `gdf.has_crs()` or something that would be preferable. But I'm not aware of such a way...

Reply:
Hm, yeah I haven't looked at it this way. So when no crs is defined, we assume that the user selected the correct crs, right?
And furthermore, we assume that there is no other cause of an ``AttributeError`` within the try statement, which I guess is reasonably since it's a small function. 

Shall I add a comment to the ``pass`` statement: 
"If the coord_gdf has no crs defined, no Error is raised and it is assumed that the user set the crs correctly"

Reply:
For the case of exp.gdf without a valid 'geometry' column, it would then be as follows:
in exp.assign_centroids() one checks that ``exp.crs`` equals ``haz.centroids.crs``, then within the u_coord.assign_centroids_to_gdf() one tries and fails to check that ``exp.gdf.crs`` equald ``haz.centroids.crs``, and it is cought by the ``except AttributeError`` statement, because exp.gdf does not have a valid geometry column.